src/bit.h

```diff
@@ -0,0 +1,48 @@
+// This file is under the public domain.
+
+#pragma once
+
+#include <cstddef>
+#ifdef _WIN32
+#include <intrin.h>
+#endif
+#include "common_types.h"
+
+namespace Common {
+
+#ifdef _MSC_VER
+static inline u8 MostSignificantSetBit(u8 val) {
+    unsigned long index;
+    _BitScanReverse(&index, val);
+    return static_cast<u8>(index);
+}
+static inline u8 MostSignificantSetBit(u16 val) {
+    unsigned long index;
+    _BitScanReverse(&index, val);
+    return static_cast<u8>(index);
+}
+static inline u8 MostSignificantSetBit(u32 val) {
+    unsigned long index;
+    _BitScanReverse(&index, val);
+    return static_cast<u8>(index);
+}
+static inline u8 MostSignificantSetBit(u64 val) {
+    unsigned long index;
+    _BitScanReverse64(&index, val);
+    return static_cast<u8>(index);
+}
+#else
+static inline u8 MostSignificantSetBit(u8 val) {
+    return static_cast<u8>(sizeof(u8) - __builtin_clz(val) - 1);
```

Comment:
Sorry that I missed this in the previous pass. This has multiple problems:
 - you need * 8 to get the bit count
 - `__builtin_clz` always accept the parameter of type `unsigned int` and hence always count the bit in the 0~31 range for x64 system despite of what value type you pass in. 

The correct way to do this is to use the same implementation `sizeof(unsigned int) * 8 - __builtin_clz(val) - 1` for all three implementations for `u8`, `u16` and `u32`. The `u64` one should be `sizeof(unsigned long long) * 8 - __builtin_clzll(val) - 1`, with a weaker assumption of `unsigned long long` >= `u64` instead of the current `unsigned long long` = `u64`